File: Bloom Filter/bf.cpp
#include <iostream>
#include <bitset>
#include <vector>
#include <string>
#include <functional>
using namespace std;
class BloomFilter
{
private:
  static const int SIZE = 1000; // Size of bit array
  bitset<SIZE> bitArray;
  vector<hash<string>> hashFunctions;

public:
  BloomFilter(int numHashes = 3)
  {
    // Initialize multiple hash functions (using std::hash with different seeds)
    for (int i = 0; i < numHashes; ++i)
    {
      hashFunctions.push_back(hash<string>());
    }
  }

  void add(const string &item)
  {
    for (size_t i = 0; i < hashFunctions.size(); ++i)
    {
      size_t hash = hashFunctions[i](item + to_string(i)) % SIZE;
      bitArray.set(hash);
    }
  }

  bool check(const string &item) const
  {
    for (size_t i = 0; i < hashFunctions.size(); ++i)
    {
      size_t hash = hashFunctions[i](item + to_string(i)) % SIZE;
      if (!bitArray.test(hash))
      {
        return false;
      }
    }
    return true;
  }
};

// 🔧 Example usage
int main()
{
  BloomFilter bf;

  bf.add("apple");
  bf.add("banana");

  cout << std::boolalpha;
  cout << "apple: " << bf.check("apple") << "\n";   // true
  cout << "banana: " << bf.check("banana") << "\n"; // true
  cout << "cherry: " << bf.check("cherry") << "\n"; // false (probably)

  return 0;
}

File: Bloom Filter/bf.py
import math
import mmh3  # MurmurHash
from bitarray import bitarray

class BloomFilter:
    def __init__(self, size: int, hash_count: int):
        self.size = size
        self.hash_count = hash_count
        self.bit_array = bitarray(size)
        self.bit_array.setall(0)

    def add(self, item: str):
        for i in range(self.hash_count):
            index = mmh3.hash(item, i) % self.size
            self.bit_array[index] = 1

    def check(self, item: str) -> bool:
        for i in range(self.hash_count):
            index = mmh3.hash(item, i) % self.size
            if self.bit_array[index] == 0:
                return False
        return True

# Example usage
bf = BloomFilter(size=5000, hash_count=7)
bf.add("apple")
bf.add("banana")

print(bf.check("apple"))   # True
print(bf.check("banana"))  # True
print(bf.check("cherry"))  # False (probably)
